docs/yum-complete-transaction.1
.\" yum-complete-transaction
.TH "yum-complete-transaction" "1" "2007 Dec 10" "Seth Vidal" ""
.SH "NAME"
yum-complete-transaction
.SH "SYNOPSIS"
\fByum-complete-transaction\fP
.SH "DESCRIPTION"
.PP 
\fByum-complete-transaction\fP is a program which finds incomplete or
aborted yum transactions on a system and attempts to complete them. It
looks at the transaction-all* and transaction-done* files which can normally
be found in /var/lib/yum if a yum transaction aborted in the middle of 
execution.
.PP
If it finds more than one unfinished transaction it will attempt to complete
the most recent one first. You can run it more than once to clean up all
unfinished transactions.
.PP 
.SH "FILES"
As yum-complete-transaction uses YUM libraries for retrieving all the information, it
relies on YUM configuration for its default values like which repositories
to use. Consult YUM documentation for details:
.PP
.nf 
/etc/yum.conf
/etc/yum/repos.d/
/var/cache/yum/
.fi 

.PP 
.SH "SEE ALSO"
.nf
.I yum.conf (5)
.\"http://linux.duke.edu/yum-utils/
http://linux.duke.edu/yum/
.fi 

.PP 
.SH "AUTHORS"
.nf 
See the Authors file included with this program.
.fi 

.PP 
.SH "BUGS"
There of course aren't any bugs, but if you find any, they should be sent
to the mailing list: yum@lists.linux.duke.edu or filed in bugzilla.
.fi


ChangeLog
2007-12-07  Tim Lauridsen <tim@naboo.local>

	* ChangeLog, yum-utils.spec: bumped yum-utils version to 1.1.9

2007-12-07  Tim Lauridsen <tim@naboo.local>

	* yum-utils.spec: bumped yum-utils version to 1.1.9

2007-12-06  Seth Vidal <skvidal@fedoraproject.org>

	* package-cleanup.py: get rid of the confirmation to clean up old
	dupes when there are no dupes to clean up. rh bug: 328711

2007-12-03  Luke Macken <lmacken@redhat.com>

	* plugins/fastestmirror/ChangeLog,
	plugins/fastestmirror/fastestmirror.conf,
	plugins/fastestmirror/fastestmirror.py: Support multiple mirror
	excludes in fastestmirror

2007-12-03  Luke Macken <lmacken@redhat.com>

	* plugins/fastestmirror/ChangeLog: update fastestmirrors ChangeLog

2007-12-03  Luke Macken <lmacken@redhat.com>

	* plugins/fastestmirror/fastestmirror.py: Add an 'exclude'
	configuration option to fastestmirror (Bug #407471)

2007-12-03  Luke Macken <lmacken@redhat.com>

	* plugins/fastestmirror/fastestmirror.conf,
	plugins/fastestmirror/fastestmirror.py: Add a maxthreads
	configuration option to fastestmirror (Bug #227772)

2007-12-03  Luke Macken <lmacken@redhat.com>

	* plugins/fastestmirror/fastestmirror.py: Actually display the
	fastest mirrors

2007-12-03  Luke Macken <lmacken@redhat.com>

	* plugins/fastestmirror/fastestmirror.py: Remove unused
	PluginYumExit import

2007-11-23  Tim Lauridsen <tla@rasmil.dk>

	* package-cleanup.py: Added patch by Ville Skytta to make
	'package-cleanup --leaves' catch *lib and *libs

2007-11-21  Seth Vidal <skvidal@fedoraproject.org>

	* yum-builddep.py: port yum-builddep to YumUtilBase format. Make it
	actually, work :) closes rh bug #381291

2007-11-21  Seth Vidal <skvidal@fedoraproject.org>

	* package-cleanup.py: clean up a deprecation

2007-11-19  Tim Lauridsen <tla@rasmil.dk>

	* plugins/basearchonly/basearchonly.conf,
	plugins/basearchonly/basearchonly.py: basearchobly patch by Adel
	Gadllah * Fix system-install-packages crash on F8 * Add wildcard
	support to whitelist * Add all wine packages to the default
	whitelist on x86 (wine*)

2007-11-12  Tim Lauridsen <tim@localhost.localdomain>

	* Makefile, plugins/priorities/ChangeLog,
	plugins/priorities/priorities.py: Added optparser check in
	priorities plugin (bz #847)

2007-11-05  Tim Lauridsen <tla@rasmil.dk>

	* plugins/priorities/ChangeLog, plugins/priorities/README,
	plugins/priorities/priorities.py: Make the name.arch checking
	optional in the priorities plugin. This adds the 'only_samearch'
	plugin option and the '--samearch-priorities'

2007-10-29  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/basearchonly/basearchonly.conf: apply patch from Adel
	Gadllah to add libflashsupport to basearchonly-plugin whitelist

2007-10-29  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py: apply patch from Bill Nottingham to make repoquery
	more silent when --quiet is specified.

2007-10-29  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/upgrade-helper/cleanup.xml,
	plugins/upgrade-helper/remove-stuff.xml: rename remove-stuff.xml to
	cleanup.xml

2007-10-26  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/upgrade-helper/upgrade-helper.py: make it work with more
	recent matchPackageNames than when I originally wrote it

2007-10-26  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: left out upgrade-helper from the list of plugins

2007-10-26  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/upgrade-helper/remove-stuff.xml,
	plugins/upgrade-helper/upgrade-helper.conf,
	plugins/upgrade-helper/upgrade-helper.py, yum-utils.spec: add
	upgrade-helper - this _should_ work - but I'll be testing it some
	more before tagging

2007-10-25  Seth Vidal <skvidal@fedoraproject.org>

	* yumdownloader.py: catch error on download if destdir is not
	writable. closed rh bug 352911

2007-10-17  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2007-10-17  Tim Lauridsen <timlau@fedoraproject.org>

	* Makefile, yum-utils.spec: bumped yum-utils version to 1.1.8

2007-10-17  Tim Lauridsen <tim@localhost.localdomain>

	* updateonboot/yum-updateonboot.init, yumdownloader.py: Make
	yumdownloader work as non-root users even if *-source repos is
	enabled in *.repo files or by --enablerepo.

2007-10-17  Seth Vidal <skvidal@fedoraproject.org>

	* yumdownloader.py: catches an error if someone has --enablerepo or
	--disablerepo set and tries to tell them what to do.

2007-10-16  Seth Vidal <skvidal@fedoraproject.org>

	* yumdownloader.py: make sure enabled source repos don't cause
	yumdownloader --source to not do the right thing

2007-10-15  Seth Vidal <skvidal@fedoraproject.org>

	* repo-rss.py: add config file option to repo-rss

2007-10-15  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'master' of
	ssh://login.linux.duke.edu/home/groups/yum/git/yum-utils  * 'master'
	of ssh://login.linux.duke.edu/home/groups/yum/git/yum-utils:
	provide more verbose failure information with higher debug level
	check for yum-updatesd being on the bus in case we're using system
	activation   refresh-updatesd: documentation updates
	refresh-updatesd: Remove repos_setup check.   refresh-updatesd: some
	tweaks   fixed issue in debuginfo-install there make it fail with
	plugins there uses conduit.getPackages() (rhbz #249742)   chmod +x
	*.py (make it a little easier to test)   fixed UnboundLocalError:
	local variable 'r_e' referenced before assignment in repoquery
	(convert unicode -> str)   Make package-cleanup --orphans work as no
	root (rhbz #253352)   Fixed yumdownloader --url output to show the
	right url (rhbz #301981)   Allow package name globs on the
	list-sec/info-sec lines   We need to catch when we install a kmod
	with a lower VR than a kmod of   Don't print metadata more than once


2007-10-15  Seth Vidal <skvidal@fedoraproject.org>

	* repo-rss.py: - apply guid patch from ricky zhou to repo-rss.xml - add -a option
	to specify arches that repo-rss should use

2007-10-09  Jeremy Katz <katzj@redhat.com>


yum-complete-transaction.py
#!/usr/bin/python -tt

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

import sys
sys.path.insert(0,'/usr/share/yum-cli')

import yum
from yum.misc import getCacheDir

from cli import *
from utils import YumUtilBase

from urlparse import urljoin
from urlgrabber.progress import TextMeter


try:
   from yum.misc import find_unfinished_transactions, find_ts_remaining
except ImportError:

    import glob
    import os
    import os.path
        
    def find_unfinished_transactions(yumlibpath='/var/lib/yum'):
        """returns a list of the timestamps from the filenames of the unfinished 
           transactions remaining in the yumlibpath specified.
        """
        timestamps = []    
        tsallg = '%s/%s' % (yumlibpath, 'transaction-all*')
        tsdoneg = '%s/%s' % (yumlibpath, 'transaction-done*')
        tsalls = glob.glob(tsallg)
        tsdones = glob.glob(tsdoneg)

        for fn in tsalls:
            trans = os.path.basename(fn)
            timestamp = trans.replace('transaction-all.','')
            timestamps.append(timestamp)

        timestamps.sort()
        return timestamps
        
    def find_ts_remaining(timestamp, yumlibpath='/var/lib/yum'):
        """this function takes the timestamp of the transaction to look at and 
           the path to the yum lib dir (defaults to /var/lib/yum)
           returns a list of tuples(action, pkgspec) for the unfinished transaction
           elements. Returns an empty list if none.

        """
        
        to_complete_items = []
        tsallpath = '%s/%s.%s' % (yumlibpath, 'transaction-all', timestamp)    
        tsdonepath = '%s/%s.%s' % (yumlibpath,'transaction-done', timestamp)
        tsdone_items = []

        if not os.path.exists(tsallpath):
            # something is wrong, here, probably need to raise _something_
            return to_complete_items    

                
        if os.path.exists(tsdonepath):
            tsdone_fo = open(tsdonepath, 'r')
            tsdone_items = tsdone_fo.readlines()
            tsdone_fo.close()     
        
        tsall_fo = open(tsallpath, 'r')
        tsall_items = tsall_fo.readlines()
        tsall_fo.close()
        
        for item in tsdone_items:
            # this probably shouldn't happen but it's worth catching anyway
            if item not in tsall_items:
                continue        
            tsall_items.remove(item)
            
        for item in tsall_items:
            item = item.replace('\n', '')
            if item == '':
                continue
            (action, pkgspec) = item.split()
            to_complete_items.append((action, pkgspec))
        
        return to_complete_items

class YumCompleteTransaction(YumUtilBase):
    NAME = 'yum-complete-transactions'
    VERSION = '1.0'
    USAGE = '"usage: yum-complete-transaction'
    
    def __init__(self):
        YumUtilBase.__init__(self,
                             YumCompleteTransaction.NAME,
                             YumCompleteTransaction.VERSION,
                             YumCompleteTransaction.USAGE)
        self.logger = logging.getLogger("yum.verbose.cli.yumcompletets")                             
        self.main()

    def clean_up_ts_files(self, timestamp, path):
    
        # clean up the transactions 
        tsdone = '%s/transaction-done.%s' % (path, timestamp)
        tsall = '%s/transaction-all.%s' % (path, timestamp)        
        os.unlink(tsdone)
        os.unlink(tsall)

    def main(self):
        # Add util commandline options to the yum-cli ones
        parser = self.getOptionParser() 
        # Parse the commandline option and setup the basics.
        try:
            opts = self.doUtilConfigSetup()
        except yum.Errors.RepoError, e:
            self.logger.error("Cannot handle specific enablerepo/disablerepo options.")
            sys.exit(50)

        # Check if there is anything to do.
#        if len(self.cmds) < 1: 
#            parser.print_help()
#            sys.exit(0)

        if self.conf.uid != 0:
            self.logger.error("Error: You must be root to finish transactions")
            sys.exit(1)

        # Setup yum (Ts, RPM db, Repo & Sack)
        self.doUtilYumSetup()
        # Do the real action
        # get the list of transactions remaining
        # list the number of them
        # take the most recent one
        # populate the ts
        # run it
        times = find_unfinished_transactions(self.conf.persistdir)
        if not times:
            print "No unfinished transactions left."      
            sys.exit()
        
        print "There are %d outstanding transactions to complete. Finishing the most recent one" % len(times)    
        
        timestamp = times[-1]
        remaining = find_ts_remaining(timestamp, yumlibpath=self.conf.persistdir)
        print "The remaining transaction had %d elements left to run" % len(remaining)
        for (action, pkgspec) in remaining:
            if action == 'install':
               try:
                   self.install(pattern=pkgspec)
               except yum.Errors.InstallError, e:
                   pass

            if action == 'erase':
               (e, m, u) = self.rpmdb.matchPackageNames([pkgspec])
               for pkg in e:
                   self.remove(po=pkg)
                   
                  

        self.buildTransaction()
        if len(self.tsInfo) < 1:
            print 'Nothing in the unfinished transaction to cleanup.'
            print "Cleaning up completed transaction file"            
            self.clean_up_ts_files(timestamp, self.conf.persistdir)
            sys.exit()
            
        else:            
            if self.doTransaction():
                print "Cleaning up completed transaction file"            
                self.clean_up_ts_files(timestamp, self.conf.persistdir)
                sys.exit()
             


        
        
    
if __name__ == '__main__':
    util = YumCompleteTransaction()
        
       
